### ISSUE TITLE
enhancement: Maintenance information

### ISSUE THREAD

**Issue by MSoeb** [OpenSlides/openslides-client#1997]
Description:
I think it would be good to have a maintenance information bar at the login page of an instance and also in the instance itself where users could get information about it. This would help users to plan or stop their work in the instance. Especially the second case could be quite annoying for active users in the instance, if a maintenance takes place and they can't save something. 

Realization:
A) Snackbar on login page with short information what happens and how long it takes. 
B) Drop down banner for user which are logged in with information about what happen and how long it takes.

Additonally: In both cases, it would be helpful if the information could be displayed beforehand so that users have the information in time.

**Comment by luisa-beerboom** [comment]
1. In which cases would the bar be used?
2. What would trigger it?
3. Will the content be generated by the instance (i.e. pre-determined) or decided by whoever is conducting the "maintenance"?
4. If it is supposed to be pre-determined, what should it say?
5. Is there supposed to be some additional functionality (since you said "plan" I would assume that the bar would perhaps need to warn the users beforehand)?

@MSoeb  I can't say anything about the implementation of this feature unless I have some more information.

**Comment by MSoeb** [comment]
1. Example Cases: A) For an Update, with a short-medium downtime. B) For a longer downtime; e.g. migration with took several hours. C) If an update took place, where the instance is still running, but the maintenance process affects the working on it. D) The instance is down. 
2. Two triggers could be possible: A) Instance is oflline without a clear reason -> bar on login page. B) Triggered by admins with information about how long it will take until the instance is online again or when it could be used again without negative effects.
3. In think, in case 2A it could be generated automatically and in case 2B an admin should write the comment based on a template.
4. Something like "Der Server ist derzeit nicht erreichbar." for case 2A. 
5. Yes indeed. i think it would be nice to have some kind of system that gave the admins the possibily to set an information bar in a running instance with tells the users when the system maintenance will take place. So it's easier for them plan their work.

**Comment by luisa-beerboom** [comment]
The way I see it this probably should simply be bar that can be customized and turned on/off by the superadmin(?). The text should be generated automatically with a couple of parameters that the admin can determine (something like `start_time`).

Text should then be something like `Starting on Monday (XX-XX-XXXX) at XX:XX the server will be undergoing maintenance and may not be available. This may take a while.` before the start date has come, and `The server is undergoing maintenance. This may take a while, please check back later.` afterwards.

@MSoeb Does this fit your requirements?

Given that this should supposedly be triggered by the one running the instance, would we want a script that can toggle this per command line? (Maybe even for multiple instances at once?)

I'm not quite sure how we should get this information in the client, I'm also not quite sure whether the client can even be accessed by the user while maintenance/update/migration is in progress.

@jsangmeister @peb-adr Do you have ideas how realistic something like this is/how something like this could function?

**Comment by peb-adr** [comment]
I think we should reduce the requirements for this a bit.

> 1. Example Cases: A) For an Update, with a short-medium downtime. B) For a longer downtime; e.g. migration with took several hours. C) If an update took place, where the instance is still running, but the maintenance process affects the working on it. D) The instance is down.
> 
>  2. Two triggers could be possible: A) Instance is oflline without a clear reason -> bar on login page. B) Triggered by admins with information about how long it will take until the instance is online again or when it could be used again without negative effects.

2A (including 1D) Is a case which IMO makes no sense to be handled by OpenSlides as a software feature. If the instance is actually unreachable the client would have no way of retrieving this info (or displaying it, except if it's cached by the browser which 
we cannot rely on). In our setup this is also already handled by surrounding infrastructure (e.g. try opening https://non-xistent.openslides.com/).

2B however is (the way I see it) the very point of this issue and a sensible feature to have - An easy way to get an arbitrary message out to every user in a well visible but not impairing manner.

My vision of how to realize this would include:
- Add a new field to the `organization` collection, e.g. `admin_message` or `broadcast_message`
  - Optionally adding e.g. `admin_message_start_time` and `admin_message_end_time` as @luisa-beerboom suggested may be desirable
  - At this point we may want to create `admin_message` as a separate collection and have a relation in `organization`
- Ensure the the suitable `update` action in the backend is compatible (depending on which of the aforementioned options we'd take)
- Add input forms in the superadmin settings to be able to set them through the web interface
- Subscribe to these fields along the rest of the organization fields in the client
  - Show the banner/snackbar when appropriate (i.e. it is set and `start_time` `end_time` align)

I personally would prefer to have this message pop up in the snack bar. It would catch the attention of the user but would be dismissable and not be in the way, which the banner could be IMO. Also I presume it would be a more l linear implementation in the client, as we'd have only one reusable way of showing it to logged-in as well as logged-out users
(This is assuming the top-banner, as used for 'offline mode' or 'new vote in progress', is not usable when logged out)
But I am willing to have my mind changed about this ...

> Given that this should supposedly be triggered by the one running the instance, would we want a script that can toggle this per command line? (Maybe even for multiple instances at once?)

This is indeed a requirement I'd bring to this discussion. With the way I described it would now already be fulfilled. Via the manage tool's `set organization` command (which internally calls the corresponding backend `action`) we'd have the ability to set it, naturally along the already existing ones.

> I'm not quite sure how we should get this information in the client, I'm also not quite sure whether the client can even be accessed by the user while maintenance/update/migration is in progress.

Availability of this information may of course differ case-by-case. However as I said before, if the instance is broken to to the point of not functioning correctly it wouldn't be sensible to rely on a software feature to convey just that.
But being able to announce the possibility of that beforehand or really any other technical Info relevant to all users is something I see as helpful and a good idea.

**Comment by jsangmeister** [comment]
I agree whole-heartedly with @peb-adr. However, I don't think a new collection is necessary for this, since it would just be a 1:1 relation and would make the creation and update more complex, so just 1-4 fields (`message`, `start_time`?, `end_time`? and maybe a boolean `show` to disable it without removing the message?) should suffice. We could also think about displaying the message in an extra section on the login page as an additional way to reach the users. Or above the calendar? (just throwing around some ideas here)
